### ISSUE TITLE
Order matters while Resolving a Union Type and its conditional keys 

### ISSUE THREAD

**Issue by azizhk** [microsoft/TypeScript#29703]
**TypeScript Version:**  3.4.0-dev.20190202

**Search Terms:** Conditional Resolution, Optionals

**Code**

```ts
interface RangeProps {
    range?: false;
    onApply?: (value: Date) => void;
}
interface SingleProps {
    range: true;
    onApply?: (value: [Date, Date]) => void;
}
type CalendarProps = RangeProps | SingleProps;

function apply (props: CalendarProps) {
    if (props.onApply) {
        if (props.range) {
            // Should work as already inside props.onApply check.
            props.onApply([new Date(), new Date()]);
        } else {
            // Should work as already inside props.onApply check.
            props.onApply(new Date());
        }
    }
}
```

**Expected behavior:**
Above code should not throw error.

**Actual behavior:**
Error at lines after comment: Cannot invoke an object which is possibly 'undefined'.

[**Playground Link:**](https://typescript-play.js.org/#code/JYOwLgpgTgZghgYwgAgEpxAcwgBSgewAcBnZAbwChlrkoNsB+ALmXgBtiIBuKm-EAIKFCbAJ7NkACgBucNgFcILACJxIASmQBeAHzJp+YABMeAXwqhIsRCgDKoTG1wES5XtTpYlyMFEU8aZH4hEXEWGTlFFgBtVUgAGmQ4iABdTV19QxMKczBRQhQAYTkIECM4KDwiUi00emdq5AAfZHssJyqSHgoYeRAEMGB+ZDhhMSlCF2IWYqcyis7iTUpA4BgJqYA6YLHRZfdA5DWN6s3PbH3Dq+RJ053QyWiQCAB3JLUISXVE57fkr7SAWupmQEA4KBW10CtxI20Eu0kv3ekC+6iBV3MgXMpiAA)

**Related Issues:** 
Tried my best, sorry but could not find any
Please also change/suggest the title of issue if it does not reflect on the error.

**Comment by RyanCavanaugh** [comment]
Basically a duplicate of #28131
